massa-event-cache/src/event_cache.rs

```diff
@@ -0,0 +1,1730 @@
+// std
+use std::cmp::max;
+use std::collections::{BTreeMap, BTreeSet};
+use std::path::Path;
+// third-party
+use num_enum::IntoPrimitive;
+use rocksdb::{IteratorMode, Options, WriteBatch, DB};
+use tracing::{debug, warn};
+// internal
+use crate::rocksdb_operator::counter_merge;
+use crate::ser_deser::{
+    SCOutputEventDeserializer, SCOutputEventDeserializerArgs, SCOutputEventSerializer,
+};
+use massa_models::address::Address;
+use massa_models::error::ModelsError;
+use massa_models::execution::EventFilter;
+use massa_models::operation::{OperationId, OperationIdSerializer};
+use massa_models::output_event::SCOutputEvent;
+use massa_models::slot::Slot;
+use massa_serialization::{DeserializeError, Deserializer, Serializer};
+
+const OPEN_ERROR: &str = "critical: rocksdb open operation failed";
+// const COUNTER_INIT_ERROR: &str = "critical: cannot init rocksdb counters";
+const DESTROY_ERROR: &str = "critical: rocksdb delete operation failed";
+const CRUD_ERROR: &str = "critical: rocksdb crud operation failed";
+const EVENT_DESER_ERROR: &str = "critical: event deserialization failed";
+const OPERATION_ID_DESER_ERROR: &str = "critical: deserialization failed for op id in rocksdb";
+const COUNTER_ERROR: &str = "critical: cannot get counter";
+const COUNTER_KEY_CREATION_ERROR: &str = "critical: cannot create counter key";
+
+#[allow(dead_code)]
+/// Prefix u8 used to identify rocksdb keys
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, IntoPrimitive)]
+#[repr(u8)]
+enum KeyIndent {
+    Counter = 0,
+    Event,
+    EmitterAddress,
+    OriginalCallerAddress,
+    OriginalOperationId,
+    IsError,
+    IsFinal,
+}
+
+/// Key type that we want to get
+enum KeyBuilderType<'a> {
+    Slot(&'a Slot),
+    Event(&'a Slot, u64),
+    Address(&'a Address),
+    OperationId(&'a OperationId),
+    Bool(bool),
+    None,
+}
+
+struct DbKeyBuilder {
+    /// Operation Id Serializer
+    op_id_ser: OperationIdSerializer,
+}
+
+impl DbKeyBuilder {
+    fn new() -> Self {
+        Self {
+            op_id_ser: OperationIdSerializer::new(),
+        }
+    }
+
+    /// Low level key builder function
+    /// There is no guarantees that the key will be unique
+    /// Recommended to use high level function like:
+    /// `key_from_event`, `prefix_key_from_indent`,
+    /// `prefix_key_from_filter_item` or `counter_key_from_filter_item`
+    fn key(
+        &self,
+        indent: &KeyIndent,
+        key_type: KeyBuilderType,
+        _is_prefix: bool,
+        is_counter: bool,
```

Comment:
`if(is_counter)` is the same as

```
if indent == KeyIndent::Counter {
  assert!(value > 40);
}
```

right ? might as well drop the is_counter boolean to reduce redundancy and error surface